lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -338,6 +347,27 @@ def write(chunk)
       # Group the entries since we have to make one call per tag.
       grouped_entries = {}
       chunk.msgpack_each do |tag, *arr|
+        tag = tag.to_s
+        if @require_valid_tags
+          # When require_valid_tags is true, invalid cases will cause the log to
+          # be dropped and warnings generated in the log.
+          if convert_to_utf8(tag) != tag || tag.match(@invalid_tag_regexp)
+            @log.warn "Dropping log message(s) with invalid tag: '#{tag}'. " \
+                      'A tag should be a string with alphanumeric characters.'
+            next
+          end
+        else
+          # When require_valid_tags is false, try to convert any invalid tag by
+          # stripping off invalid characters.
+          converted_tag = convert_to_utf8(tag, '').gsub(@invalid_tag_regexp, '')
```

Comment:
Spaces in tags are not allowed by the Logging API. The only characters other than alphanumeric ones that are allowed are: `-`, `_`, `/` and `.`

Our previous implementation (replacing non-utf8 characters with spaces in tags) would actually fail the Logging API requests. The user configured replacement string can contain characters other than this set, thus we can not safely use that.

Potentially we can hard code a string (only includes alphanumeric, `-`, `_`, `/` and `.`) to replace invalid characters with. Or we can add a separate config if we want the users to be able to configure this string, but make them choose from these characters.

Thoughts?